test/functional/rpc_bind.py

```diff
@@ -89,19 +83,26 @@ def run_test(self):
         # check only IPv4 localhost (explicit) with multiple alternative ports on same host
         self.run_bind_test(['127.0.0.1'], '127.0.0.1:32171', ['127.0.0.1:32171', '127.0.0.1:32172'],
             [('127.0.0.1', 32171), ('127.0.0.1', 32172)])
-        # check only IPv6 localhost (explicit)
-        self.run_bind_test(['[::1]'], '[::1]', ['[::1]'],
-            [('::1', defaultport)])
-        # check both IPv4 and IPv6 localhost (explicit)
-        self.run_bind_test(['127.0.0.1'], '127.0.0.1', ['127.0.0.1', '[::1]'],
-            [('127.0.0.1', defaultport), ('::1', defaultport)])
-        # check only non-loopback interface
-        self.run_bind_test([non_loopback_ip], non_loopback_ip, [non_loopback_ip],
-            [(non_loopback_ip, defaultport)])
+        if have_ipv6:
+            self.log.info("Using interface %s for testing" % non_loopback_ip)
+
+            # check only IPv6 localhost (explicit)
+            self.run_bind_test(['[::1]'], '[::1]', ['[::1]'],
+                [('::1', defaultport)])
+            # check both IPv4 and IPv6 localhost (explicit)
+            self.run_bind_test(['127.0.0.1'], '127.0.0.1', ['127.0.0.1', '[::1]'],
+                [('127.0.0.1', defaultport), ('::1', defaultport)])
+
+        if non_loopback_ip is not None:
+            # check only non-loopback interface
+            self.run_bind_test([non_loopback_ip], non_loopback_ip, [non_loopback_ip],
+                [(non_loopback_ip, defaultport)])
 
-        # Check that with invalid rpcallowip, we are denied
-        self.run_allowip_test([non_loopback_ip], non_loopback_ip, defaultport)
-        assert_raises_rpc_error(-342, "non-JSON HTTP response with '403 Forbidden' from server", self.run_allowip_test, ['1.1.1.1'], non_loopback_ip, defaultport)
+            # Check that with invalid rpcallowip, we are denied
+            self.run_allowip_test([non_loopback_ip], non_loopback_ip, defaultport)
+            assert_raises_rpc_error(-342, "non-JSON HTTP response with '403 Forbidden' from server", self.run_allowip_test, ['1.1.1.1'], non_loopback_ip, defaultport)
```

Comment:
The allowip test could also be done by connecting from a second loopback IP, say, 127.0.0.2. This would obliterate the need for a non loopback IP, but would need support at the RPC/HTTP level for binding to a different address at connection (connecting to a different interface IP does that part automatically...)

Reply:
Ughh, I tried this but it doesn't work. Apparently the entire IPv4 subnet is unconditionally allowed.
```c++
rpc_allow_subnets.push_back(CSubNet(localv4, 8));      // always allow IPv4 local subnet
```

Reply:
@laanwj It depends on the IP stack implementation.  Some only allow the use of only 127.0.0.1 while others seem to effectively bind 127/8.  I think it would be best to rely only on 127.0.0.1 and ::1 for loopback-based IPv4/IPv6 compatibility checks.

Reply:
This is a linux-specific test though, so in this test we can asume `/8` and don't care about other IP stack implementations.